modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java

```diff
@@ -0,0 +1,263 @@
+
+            tableManager.tableAsyncInternal(canonicalName).thenAccept((table) -> {
+                if (table == null) {
+                    var exception = new TableNotFoundException(canonicalName);
+
+                    LOG.info("Unable to create index [schemaName={}, tableName={}, indexName={}]",
+                            exception, schemaName, tableName, indexName);
+
+                    future.completeExceptionally(exception);
+
+                    return;
+                }
+
+                tableManager.alterTableAsync(table.name(), tableChange -> tableChange.changeIndices(indexListChange -> {
+                    if (indexListChange.get(indexName) != null) {
+                        var exception = new IndexAlreadyExistsException(indexName);
+
+                        LOG.info("Unable to create index [schemaName={}, tableName={}, indexName={}]",
+                                exception, schemaName, tableName, indexName);
+
+                        future.completeExceptionally(exception);
+
+                        return;
+                    }
+
+                    indexListChange.create(indexName, indexChange);
+
+                    TableIndexView indexView = indexListChange.get(indexName);
+
+                    Set<String> columnNames = Set.copyOf(tableChange.columns().namedListKeys());
+
+                    validateColumns(indexView, columnNames);
+                })).whenComplete((index, th) -> {
+                    if (th != null) {
+                        LOG.info("Unable to create index [schemaName={}, tableName={}, indexName={}]",
+                                th, schemaName, tableName, indexName);
+
+                        future.completeExceptionally(th);
+                    } else if (!future.isDone()) {
+                        // don't pay too much attention to this part because it should be reworked
+                        // within refactoring of schema objects: when both tables and indexes will be derived
+                        // from common parent and become entity of the same level in the configuration,
+                        // then there sill be no need to look up indices in such way
+
+                        var createdIndex = indices.values().stream()
+                                .filter(idx -> indexName.equals(idx.name()))
+                                .findFirst()
+                                .orElse(null);
+
+                        if (createdIndex != null) {
+                            LOG.info("Index created [schemaName={}, tableName={}, indexName={}, indexId={}]",
+                                    schemaName, tableName, indexName, createdIndex.id());
+
+                            future.complete(createdIndex);
+                        } else {
+                            var exception = new IgniteInternalException(
+                                    Common.UNEXPECTED_ERR, "Looks like the index was concurrently deleted");
```

Comment:
is it possible? I thought that all changer run sequentially and can't be run concurrently

Reply:
>  I thought that all changer run sequentially and can't be run concurrently

This is true. But the line you are pointing to is not a part of a changer closure. This actually a callback that would be invoked _after_ the changer closure is applied to the configuration